test/e2e/app-dir/rewrites-redirects/rewrites-redirects.test.ts

```diff
@@ -0,0 +1,87 @@
+import { NextInstance } from 'test/lib/next-modes/base'
+import webdriver from 'next-webdriver'
+import { waitFor } from 'next-test-utils'
+
+describe('redirects and rewrites', () => {
+  let next: NextInstance
+
+  beforeAll(async () => {
+    next = await createNext({
+      files: new FileRef(__dirname),
+      dependencies: {
+        react: 'latest',
+        'react-dom': 'latest',
+        typescript: 'latest',
+        '@types/react': 'latest',
+        '@types/node': 'latest',
+      },
+    })
+  })
+  afterAll(() => next.destroy())
+
+  /**
+   * All test will use a link/button to navigate to '/*-before' which should be redirected by correct redirect/rewrite to '/*-after'
+   */
+  describe.each(['link', 'button'])('navigation using %s', (testType) => {
+    it('should rewrite from middleware correctly', async () => {
+      const browser = await webdriver(next.url, '/')
+      browser.elementById(`${testType}-middleware-rewrite`).click()
+      await waitFor(200)
+
+      expect(await browser.elementById('page').text()).toBe(
+        'middleware-rewrite-after'
+      )
+      const url = new URL(await browser.url())
+      expect(url.pathname).toEndWith('-before')
+    })
+
+    it('should redirect from middleware correctly', async () => {
+      const browser = await webdriver(next.url, '/')
+      browser.elementById(`${testType}-middleware-redirect`).click()
+      await waitFor(200)
+
+      expect(await browser.elementById('page').text()).toBe(
+        'middleware-redirect-after'
+      )
+      const url = new URL(await browser.url())
+      expect(url.pathname).toEndWith('-after')
+    })
+
+    it('should rewrite from next.config.js correctly', async () => {
+      const browser = await webdriver(next.url, '/')
+      browser.elementById(`${testType}-config-rewrite`).click()
+      await waitFor(200)
+
+      expect(await browser.elementById('page').text()).toBe(
+        'config-rewrite-after'
+      )
+      const url = new URL(await browser.url())
+      expect(url.pathname).toEndWith('-before')
+    })
+
+    it('should redirect from next.config.js correctly', async () => {
+      const browser = await webdriver(next.url, '/')
+      browser.elementById(`${testType}-config-redirect`).click()
+      await waitFor(200)
+
+      expect(await browser.elementById('page').text()).toBe(
+        'config-redirect-after'
+      )
+      const url = new URL(await browser.url())
+      expect(url.pathname).toEndWith('-after')
+    })
+
+    it('should redirect using catchall from next.config.js correctly', async () => {
+      const browser = await webdriver(next.url, '/')
+      browser.elementById(`${testType}-config-redirect-catchall`).click()
+      await waitFor(200)
+
+      expect(await browser.elementById('page').text()).toBe(
+        'config-redirect-catchall-after/thing'
+      )
+      const url = new URL(await browser.url())
+      expect(url.pathname).toEndWith('-after/thing')
+    })
```

Comment:
@OlegLustenko could you try tweaking my test to repro this?
This test is passing and I'm not sure how to repro this.
Just tweak it and then run it with 
```bash
pnpm build && pnpm testheadless test/e2e/app-dir/rewrites-redirects/rewrites-redirects.test.ts
```

Reply:
Or maybe even better would be creating a new PR with a whole new e2e test created from the original reproduction.

Reply:
I'm not sure I have access to `import { createNext, FileRef } from 'e2e-utils'`

But I would check if `createNext` is initializing the production version of nextApp

Reply:
It's running dev, production and we are deploying it to Vercel.
You will need to clone the `next.js` repo and follow the steps here in our `contribution.md` to run the tests.